cfi_backward.adoc

```diff
@@ -799,24 +780,26 @@ Svpbmt and Svnapot extensions are supported for shadow stack pages.
 
 [NOTE]
 ====
+All instructions that load from memory are allowed to read the shadow stack. The
+shadow stack only holds a copy of the link register as saved on the regular
+stack. The ability to read the shadow stack is useful for debugging, performance
+profiling, and other use cases.
+
 Operating systems should protect against writable non-shadow-stack alias
 virtual-addresses mappings being created to the physical memory of the
 shadow stack.
-====
 
-[NOTE]
-====
 Shadow stacks are expected to be bounded on each end using guard pages, so that
 no two shadow stacks are adjacent to each other. This guards against accidentally
 underflowing or overflowing from one shadow stack to another. Traditionally,
 a guard page for a stack is a page that is inaccessible to the process owning
 the stack. For shadow stacks, the guard page may also be a non-shadow-stack
 page that is otherwise accessible to the process owning the shadow stack
-because shadow stack loads and stores to non-shadow-stack pages will result in
-an exception.
+because shadow stack loads and stores to non-shadow-stack pages cause an
+access-fault exception.
```

Comment:
This choice was made to convey the fatality of this condition and simplify the page fault handler. There is text in the virtual memory system section of the spec to explain this. A page faults would require the page fault handler to read the opcode bytes from memory to make the determination whether this is a non-fixable fatal condition (non-SS store to SS memory, SS load/store to non-SS memory) or a fixable condition (V=0 or insufficient permissions). Page fault handlers are performance critical and would not want decode instructions.

Reply:
It's not always a fatal condition - if a shadow stack page needs to be made read-only (due to fork or for more esoteric reasons), a supervisor can either set V=0 (gives a page fault, but interferes with non-shadow stack read access to the page; the type of page-fault exception generated for shadow stack operations on a V=0 page needs to be documented) or change the page to rwx=100b (minimal impact on page users, but requires access fault handling).

The base privileged spec has the property that for any page fault, the requirements that the instruction places on the page permissions are always encoded in the cause value; an instruction page fault requires --X, R-X, or RWX, a store/AMO page fault requires RW- or RWX, a load page fault requires R--, RW-, R-X, RWX (or --X if sstatus.MXR=1) (or -W- if BCFIE=1).  Since all of the necessary information is in scause, page fault handlers do not need to decode instructions.

Generating load page fault exceptions from `ssload`/`sspopchk` or store/AMO page fault exceptions from `sspush`/`ssamoswap` would be incorrect since those instructions do not have the same allowed page types as base loads and stores.  The appropriate "shadow stack operations generate page faults" variant to compare to would be one where a new mcause value is allocated for "shadow stack page fault" (consistency suggests "shadow stack address fault", which would be somewhat useful if a trap handler is used to extend the number of PMP regions, and "shadow stack misaligned", which would be almost completely useless), and the property of the cause value determining the needed properties of the page without reference to the instruction is maintained.  In this case, there is no harm in generating store/AMO page faults for base store instructions to shadow stack pages: the trap handler can see from the page permissions that the page is a shadow stack page and from scause that the access is a base store/AMO and therefore cannot succeed on a shadow stack page.

Reply:
This is discussed in the virtual memory system section. In step 3, if `pte.v=0`, or any bits or encodings reserved for future standard use are set within the `pte` then a page-fault occurs. There were three options discussed with the ARC a) use access-violation b) new cause code c) new status to differentiate not-present vs. permission faults. The option c) was considered least attractive. b) could be done but was not considered necessary and a) was preferred.

Reply:
Ugly but likely tolerable and I see the benefit of not needing a cause code/bit in medeleg.  I'll spend more time thinking about whether this has subtle problems.

At a minimum, we need to specify which of the three page fault cause codes is generated by shadow stack instructions to a page with `pte.v=0` or reserved for standard use bits set.